### ISSUE TITLE
Crash during run of dotcover files

### ISSUE THREAD

**Issue by OsirisTerje** [GeorgDangl/CoberturaConverter#1]
Hi !
I've tried to use the converter with a dotcover file, and is rewarded by the following crash:
![image](https://user-images.githubusercontent.com/203432/40589852-69e24eec-61f5-11e8-87c9-41181f8e23d5.png)

I've compared the dotcover file I am using with your testfile, and they seem to have the same structure.  I have also tried both the xml and detailedxml options in DotCover, and both fail the same way. 

I then tried to clone your repo, and build it, but are then stopped by
 
![image](https://user-images.githubusercontent.com/203432/40589877-ad94d614-61f5-11e8-8e40-80c3193c93d5.png)

**Comment by GeorgDangl** [comment]
Hi Terje, thanks for the info!

I'll look into it ASAP. Can you tell me which Visual Studio version you're using?

You probably can't share the dotCover Xml report, can you?

**Comment by OsirisTerje** [comment]
Hi!  I am using VS 2017 Update 15.7.1, and no I cant share the xml unfortunately.  But, I did add it to the unit test resources you have (after trimming down the project so as to build, just removed the dot core and ran it as net471 only), and it parsed ok.   The stack trace indicates it fails in there, but as it passes the unit test, I think there is something else going on.  The file is approx 5MB, with 51000 lines. 

I just added the core package to my solution, instead of calling it as a cli, and I get the same error.  I am then calling the ConvertDotCoverToCobertura  method, and just using it as is done on the cli code

**Comment by GeorgDangl** [comment]
Unfortunately, I couldn't reproduce the build error. I checked out a fresh copy and it built as expected on Visual Studio 2017 15.7.2. As a workaround, you could deactivate the `Nuke.CoberturaConverter` package, since it's not required for the CLI build.
Could you maybe post the full output from the build log to further troubleshoot this?

As for the error itself, I think I fixed it in 1454f52. You can download the latest prerelease here to check it out:
https://www.myget.org/feed/dangl/package/nuget/CoberturaConverter.CommandLine
It's currently on version 1.1.1.-beta0002

The converter was extracting MethodInfos from dotCover and expected them to be in the format:  
`MethodName()`  
If there were no parentheses, it threw the error you encountered.

**Comment by OsirisTerje** [comment]
Ahh, cool , thanks!  I'll try that

**Comment by GeorgDangl** [comment]
Oh, that was timing=)
Thanks for the info.

Please tell me if it's working.

**Comment by OsirisTerje** [comment]
About the build issue, I'll come back to you on that, (I did strip out the nuke stuff yes, and have it building as   net471 only now).

**Comment by GeorgDangl** [comment]
You could also try to run `powershell ./build.ps1` in the root directory of the solution. It will download the dotnet CLI in a temp folder and then execute the build script. This could help in identifying the build issue.

**Comment by OsirisTerje** [comment]
I do have some managed C++ code results in there, can that be an issue - look at the ctor/Dtor syntax.
(I cant see any methods with the ()  )


Like:
```
<Type Name="StringConvert" CoveredStatements="0" TotalStatements="13" CoveragePercent="0">
        <Constructor Name="StringConvert()" CoveredStatements="0" TotalStatements="1" CoveragePercent="0" />
        <Method Name="ManagedToUnmanaged(basic_string&lt;char,std::char_traits&lt;char&gt;,std::allocator&lt;char&gt; &gt;*,string):basic_string&lt;char,std::char_traits&lt;char&gt;,std::allocator&lt;char&gt; &gt;*" CoveredStatements="0" TotalStatements="8" CoveragePercent="0" />
        <Method Name="UnmanagedToManaged(basic_string&lt;char,std::char_traits&lt;char&gt;,std::allocator&lt;char&gt; &gt;*):string" CoveredStatements="0" TotalStatements="4" CoveragePercent="0" />
      </Type>
      <Type Name="TransverseMercator" CoveredStatements="0" TotalStatements="49" CoveragePercent="0">
        <Constructor Name="TransverseMercator()" CoveredStatements="0" TotalStatements="8" CoveragePercent="0" />
        <Constructor Name="TransverseMercator(double,double,double)" CoveredStatements="0" TotalStatements="12" CoveragePercent="0" />
        <Property Name="CentralScale:double" CoveredStatements="0" TotalStatements="1" CoveragePercent="0">
          <PropertyGetter CoveredStatements="0" TotalStatements="1" CoveragePercent="0" />
        </Property>
        <Property Name="Flattening:double" CoveredStatements="0" TotalStatements="1" CoveragePercent="0">
          <PropertyGetter CoveredStatements="0" TotalStatements="1" CoveragePercent="0" />
        </Property>
        <Property Name="MajorRadius:double" CoveredStatements="0" TotalStatements="1" CoveragePercent="0">
          <PropertyGetter CoveredStatements="0" TotalStatements="1" CoveragePercent="0" />
        </Property>
        <Method Name="!TransverseMercator():void" CoveredStatements="0" TotalStatements="4" CoveragePercent="0" />
        <Method Name="~TransverseMercator():void" CoveredStatements="0" TotalStatements="1" CoveragePercent="0" />
        <Method Name="Dispose(bool):void" CoveredStatements="0" TotalStatements="1" CoveragePercent="0" />
        <Method Name="Forward(double,double,double,out double,out double):void" CoveredStatements="0" TotalStatements="4" CoveragePercent="0" />
        <Method Name="Forward(double,double,double,out double,out double,out double,out double):void" CoveredStatements="0" TotalStatements="6" CoveragePercent="0" />
        <Method Name="Reverse(double,double,double,out double,out double):void" CoveredStatements="0" TotalStatements="4" CoveragePercent="0" />
        <Method Name="Reverse(double,double,double,out double,out double,out double,out double):void" CoveredStatements="0" TotalStatements="6" CoveragePercent="0" />
      </Type>

```

And some properties like:

```
 <Type Name="DistanceBearingElevation" CoveredStatements="0" TotalStatements="11" CoveragePercent="0">
        <Constructor Name="DistanceBearingElevation(double,double,double)" CoveredStatements="0" TotalStatements="5" CoveragePercent="0" />
        <AutoProperty Name="Bearing:double" CoveredStatements="0" TotalStatements="2" CoveragePercent="0">
          <PropertyGetter CoveredStatements="0" TotalStatements="1" CoveragePercent="0" />
          <PropertySetter CoveredStatements="0" TotalStatements="1" CoveragePercent="0" />
        </AutoProperty>
        <AutoProperty Name="Distance:double" CoveredStatements="0" TotalStatements="2" CoveragePercent="0">
          <PropertyGetter CoveredStatements="0" TotalStatements="1" CoveragePercent="0" />
          <PropertySetter CoveredStatements="0" TotalStatements="1" CoveragePercent="0" />
        </AutoProperty>
        <AutoProperty Name="Elevation:double" CoveredStatements="0" TotalStatements="2" CoveragePercent="0">
          <PropertyGetter CoveredStatements="0" TotalStatements="1" CoveragePercent="0" />
          <PropertySetter CoveredStatements="0" TotalStatements="1" CoveragePercent="0" />
        </AutoProperty>
      </Type>
```

**Comment by GeorgDangl** [comment]
May I add this snippet to the repo in the test project?

**Comment by OsirisTerje** [comment]
Yes, no problem.

**Comment by OsirisTerje** [comment]
But I did add the whole file to the test, and it passed there, that was the latest code though:
![image](https://user-images.githubusercontent.com/203432/40603215-9e0e7270-625b-11e8-91c8-e31f78633723.png)


And, the myget feed link above doesn't seem to be public, is it ?

**Comment by OsirisTerje** [comment]
About the building, I went back to the developer branch, and ran : powershell .\build.ps1, and got this:

![image](https://user-images.githubusercontent.com/203432/40603643-cb0e7fbc-625c-11e8-85f8-2faa50a012e5.png)

**Comment by GeorgDangl** [comment]
Ooooh, that one is on me.

I'm referencing a version of xUnit that is no longer on their official MyGet branch. It works for me since I've got a ProGet server running that caches packages I used to access.

I'll update xUnit=)

**Comment by GeorgDangl** [comment]
And I just added a regression test, it works now but failed with the error you encountered without the fix. I'll prepare a release soon.

**Comment by GeorgDangl** [comment]
I've released v1.1.1:
https://github.com/GeorgDangl/CoberturaConverter/releases/tag/v1.1.1

It'll take a few minutes until its visible on NuGet.

Thanks a lot for the great help!

**Comment by OsirisTerje** [comment]
Cool!  Thanks for the quick response !

Btw, just debugging (before your fix), and see it crashes in GetMethodInfoFromDotCoverMethod  on a method name like:
`Attribute<MyCompany::Simplifiable>():Simplifiable`

Is this what you have fixed ?

**Comment by GeorgDangl** [comment]
Yep, that's what I fixed.

            var methodDefinition = dotCoverMethodName
                .Split(':')
                .First();
            var delimiterIndex = methodDefinition.IndexOf('(');
            if (delimiterIndex < 0)
            {
                // The delimiter index should indicate at which position
                // the opening bracket of the method starts, so that the
                // method info can be split into name an signature, e.g.:
                // name: 'GetMethodInfoFromDotCoverMethod'
                // signature: '(string dotCoverMethodName)'
                delimiterIndex = methodDefinition.Length;
            }

But looks like I need to fix it even more, since I didn't account for double `::`. Never worked with C++ before=)

**Comment by OsirisTerje** [comment]
There can be multiple of those double :: there.....

**Comment by GeorgDangl** [comment]
Fixed it:

```csharp
            // The regex should only match on single but not on double colons, eg it should match:
            // MethodName():void -> "MethodName()" and "void"
            // but not for double colons:
            // MethodName(Foo::Bar):void -> "MethodName(Foo::Bar)" and "void"
            var methodDefinition = Regex.Split(dotCoverMethodName, "(?<!:):(?!:)")
                .First();

            var delimiterIndex = methodDefinition.IndexOf('(');
            if (delimiterIndex < 0)
            {
                // The delimiter index should indicate at which position
                // the opening bracket of the method starts, so that the
                // method info can be split into name an signature, e.g.:
                // name: 'GetMethodInfoFromDotCoverMethod'
                // signature: '(string dotCoverMethodName)'
                delimiterIndex = methodDefinition.Length;
            }
            var name = methodDefinition.Substring(0, delimiterIndex);
            var signature = methodDefinition.Substring(delimiterIndex);
            return (name, signature);
```

`MethodName(Foo::Bar):void` returns a method name `MethodName` with signature `(Foo::Bar)`  
`MethodName(Foo::Bar, Fizz::Buzz):void` returns a method name `MethodName` with signature `(Foo::Bar, Fizz::Buzz)`

Is that the correct syntax?

**Comment by OsirisTerje** [comment]
Yes, but I'll add some more examples below, from the xml though:

```
<Method Name="Native&lt;MyCompany::Framework::Mechanics::ISimplifiable&gt;():ISimplifiable*" CoveredStatements="0" TotalStatements="2" CoveragePercent="0" />

<Constructor Name="DynParConnector&lt;MyCompany::MyProject::Lsd::IDynamicParameter&lt;bool&gt; &gt;(IDynamicParameter,string,bool)" CoveredStatements="0" TotalStatements="12" CoveragePercent="0" />
      
<Method Name="MyCompany.Plankton.ref_ptr&lt;MyCompany::Framework::DTO::IMechanicalCompositionDTO::ConstraintDTO&gt;.addReference(ref_ptr&lt;MyCompany::Framework::DTO::IMechanicalCompositionDTO::ConstraintDTO&gt;*):void" CoveredStatements="0" TotalStatements="4" CoveragePercent="0" />
    
```

But I think the fix you did above should handle these, as far as I can see .....

**Comment by OsirisTerje** [comment]
I checked your code above, and can confirm it now runs without any crash :-)

**Comment by GeorgDangl** [comment]
I just released v1.1.2, should be available in a few minutes on NuGet  
https://github.com/GeorgDangl/CoberturaConverter/releases/tag/v1.1.2

**Comment by OsirisTerje** [comment]
Thanks!!
